specreduce/extract.py

```diff
@@ -143,3 +145,108 @@ def _ap_weight_image(trace, width, disp_axis, crossdisp_axis, image_shape):
                           flux=ext1d * getattr(image, 'unit', u.DN))
 
         return spec
+
+
+@dataclass
+class HorneExtract(SpecreduceOperation):
+    """
+    Perform a Horne extraction (a.k.a. optimal) on a region in a
+    two-dimensional spectrum.
+    """
+
+    def __call__(self, image, variance, trace_object,
+                 disp_axis=1, crossdisp_axis=0,
+                 bkgrd_prof=models.Polynomial1D(2)):
+        """
+        Run the Horne calculation on a region of an image and extract a
+        1D spectrum.
+
+        Parameters
+        ----------
+
+        image : `~astropy.nddata.CCDData` or array-like, required
+            The input 2D spectrum from which to extract a source.
+
+        variance : `~astropy.nddata.CCDData` or array-like, required
```

Comment:
also note that in DRAGONS if no variance is available, a uniform variance is calculated based on pixel-to-pixel variations in the data. a similar fall-back could be implemented here.

Reply:
Understood on the benefits of `CCDData`. An option could be to include variance and mask keyword arguments that default to None for users who prefer to work with separate numpy arrays for whatever reason. Solely handling `CCDData` objects could move the barrier to entry higher for users depending on how familiar they are with them.

I am hesitant on calculating a variance for the user. My impression is that the calculation varies based on the instrument and what other data products you have available (weights, etc.). If that's the case, it could be hard to do that in a general manner.

Reply:
i guess an easy-ish way to do it would be to add a check if `image` is not `CCDData`, and if that's the case, make `CCDData` instance from that plus `mask` and `variance`. and we'd need a way to input a `unit` if `image` is not a `Quantity`. still, that's extra code and logic for us to maintain that wouldn't be required if we just consistently use `CCDData` internally. users that have normal numpy arrays can always just pack them into `CCDData` at call-time. e.g.:
```
horne_extract(CCDData(image, variance=variance, mask=mask, unit=unit), trace)
```
users of ground-based data that do initial processing using `ccdproc` will already have `CCDData` instances ready to go.

hear you on making variance assumptions. requiring it to be specified is a valid approach.

Reply:
If there is no variance data, I think setting it to unity (or any other constant)  everywhere will work, without having to compute the variance from the science data.

Reply:
While it would be good to support CCDData, I think we also need to keep the option to support less structured inputs, such as bare numpy arrays.  Not everyone in the community uses the CCDData format.   @eteq  What do you think?